File: app/(Pages)/News/algo.py
from django.utils import timezone
from datetime import timedelta
from django.http import JsonResponse
from .models import Interaction, Post

def get_user_feed(request, user_id):
    one_week_ago = timezone.now() - timedelta(days=7)
    interactions = Interaction.objects.filter(user_id=user_id)
    post_scores = {}

    for interaction in interactions:
        time_diff = (timezone.now() - interaction.timestamp).days
        interaction_score = max(0, 10 - time_diff)

        if interaction.type in ['like', 'share']:
            interaction_score *= 2

        related_posts = Post.objects.filter(entity_id=interaction.entity_id)
        for post in related_posts:
            post_scores[post.id] = post_scores.get(post.id, 0) + interaction_score

    # Fetch recent posts to ensure feed freshness
    recent_posts = Post.objects.filter(created_at__gte=one_week_ago)
    for post in recent_posts:
        post_scores[post.id] = post_scores.get(post.id, 5)

    # Convert to list and sort
    feed_items = [{'post': Post.objects.get(id=post_id), 'score': score}
                  for post_id, score in post_scores.items()]
    feed_items.sort(key=lambda x: x['score'], reverse=True)

    # Convert to JSON
    feed_data = [{'title': item['post'].title, 'content': item['post'].content, 'score': item['score']}
                 for item in feed_items]

    return JsonResponse(feed_data, safe=False)

File: app/(Pages)/News/algo.ts
// Assume these imports
import { Post, User, Interaction } from './models'; // Import your models

interface FeedItem {
  post: Post;
  score: number;
}

function generateUserFeed(user: User): FeedItem[] {
  const interactions = getInteractionsForUser(user.id);
  const postScores = new Map<string, number>();

  interactions.forEach(interaction => {
    // Higher score for recent interactions
    let interactionScore = Math.max(0, 10 - (new Date().getTime() - new Date(interaction.timestamp).getTime()) / (1000 * 3600 * 24));
    
    if (interaction.type === 'like' || interaction.type === 'share') {
      interactionScore *= 2; // Boost for likes and shares
    }

    const posts = getPostsByEntity(interaction.entityId); // Function to get posts by entity
    posts.forEach(post => {
      const currentScore = postScores.get(post.id) || 0;
      postScores.set(post.id, currentScore + interactionScore);
    });
  });

  // Add recent posts to ensure feed freshness
  const recentPosts = getRecentPosts();
  recentPosts.forEach(post => {
    const currentScore = postScores.get(post.id) || 0;
    postScores.set(post.id, currentScore + 5); // Recent posts get a base score
  });

  const feedItems: FeedItem[] = Array.from(postScores, ([postId, score]) => ({
    post: getPostById(postId),
    score: score
  }));

  // Sort feed items by score in descending order
  return feedItems.sort((a, b) => b.score - a.score);
}

function getInteractionsForUser(userId: string): Interaction[] {
  // Implementation to fetch user interactions from the database
}

function getPostsByEntity(entityId: string): Post[] {
  // Implementation to fetch posts from the database based on entityId
}

function getRecentPosts(): Post[] {
  // Fetch recent posts, e.g., from the last 7 days
}

function getPostById(postId: string): Post {
  // Fetch a single post by ID
}
